core/src/main/java/com/linecorp/armeria/client/ArrayBasedEventLoopState.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2023 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import io.netty.channel.EventLoop;
+
+final class ArrayBasedEventLoopState extends AbstractEventLoopState {
+
+    private final List<AbstractEventLoopEntry> entries = new ArrayList<>();
+    private final int maxNumEventLoops;
+
+    private EventLoopAcquisitionIndex acquisitionIndex;
+    private int nextVisitIndex;
+    private int allActiveRequests;
+
+    ArrayBasedEventLoopState(List<EventLoop> eventLoops, int maxNumEventLoops,
+                             DefaultEventLoopScheduler scheduler) {
+        super(eventLoops, scheduler);
+        this.maxNumEventLoops = maxNumEventLoops;
+        if (eventLoops.size() == maxNumEventLoops) {
+            // We use all event loops so initialize early.
+            init(0);
+        }
+    }
+
+    private void init(int acquisitionStartIndex) {
+        acquisitionIndex = new EventLoopAcquisitionIndex(
+                acquisitionStartIndex, maxNumEventLoops, eventLoops().size());
+        nextVisitIndex = 0;
+        addUnusedEventLoop();
+    }
+
+    private boolean addUnusedEventLoop() {
+        if (entries.size() < maxNumEventLoops) {
+            push(new Entry(this, eventLoops().get(acquisitionIndex.nextAcquirableIndex()), entries.size()));
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    AbstractEventLoopEntry acquire() {
+        lock();
+        try {
+            if (acquisitionIndex == null) {
+                init(scheduler().acquisitionStartIndex(maxNumEventLoops));
+            }
+
+            for (int offset = 0; offset < entries.size(); offset++) {
+                final int index = (nextVisitIndex + offset) % entries.size();
+                final AbstractEventLoopEntry e = entries.get(index);
+                if (e.activeRequests() == 0) {
+                    nextVisitIndex = index;
```

Comment:
Do we need `nextVisitIndex`?

How about just looping over the list while holding the index of the entry that has the lowest active requests?
```java
int minActiveRequests = Integer.MAX_VALUE;
int minIndex = 0;
for (int i = 0; i < entries.size(); i++) {
    final AbstractEventLoopEntry entry = entries.get(i);
    if (entry.activeRequests() == 0) {
        return entry;
    }
    if (minActiveRequests > entry.activeRequests()) {
        minActiveRequests = entry.activeRequests();
        minIndex = i;
    }
}
if (addUnusedEventLoop()) {
    final AbstractEventLoopEntry e = entries.get(nextVisitIndex);
    assert e.activeRequests() == 0;
}
return entries.get(minIndex);
```

Reply:
Also, we have food for thought.
Should add the entry to the list when `acquire` is called or add entries all at once when the instance is created?